### ISSUE TITLE
[deepseek_v4] save_pretrained silently downcasts FP32 tensors to BF16 (hc_*, attn_sink, ffn.gate.bias, compressor.ape, indexer.compressor.ape)

### ISSUE THREAD

**Issue by pasta-paul** [huggingface/transformers#46167]
## Summary

`save_pretrained` silently downcasts 417 FP32 tensors to BF16 when saving a DeepSeek-V4 model loaded with `torch_dtype=torch.bfloat16`. No warning, no error. The downcast loses numerical precision on plumbing tensors that DeepSeek's release spec keeps at FP32 for a reason.

## Affected tensor groups

Per DeepSeek's release at `deepseek-ai/DeepSeek-V4-Flash`, the following tensor groups are FP32 in the source safetensors:

| Pattern | Count per model | Role |
|---|---|---|
| `layers.X.hc_attn_{base,fn,scale}` + `mtp.0.hc_attn_{base,fn,scale}` | 3 × 44 = 132 | Hyper-connection attention plumbing |
| `layers.X.hc_ffn_{base,fn,scale}` + `mtp.0.hc_ffn_{base,fn,scale}` | 3 × 44 = 132 | Hyper-connection FFN plumbing |
| `model.hc_head_{base,fn,scale}` + `mtp.0.hc_head_{base,fn,scale}` | 6 | Top-level + MTP hc_head |
| `layers.X.attn.attn_sink` + `mtp.0.attn.attn_sink` | ~44 | Attention sink tokens |
| `layers.X.ffn.gate.bias` (was `e_score_correction_bias`) | 41 | MoE routing bias |
| `layers.X.attn.compressor.ape` (was `position_bias`) | 41 | Compressor positional encoding |
| `layers.X.attn.indexer.compressor.ape` (was `position_bias`) | 21 | Indexer positional encoding |
| **Total** | **417** | |

All 417 are saved as BF16 by `save_pretrained` when the model dtype is BF16.

## Repro (minimal)

```python
from transformers import AutoModelForCausalLM
import torch, safetensors
model = AutoModelForCausalLM.from_pretrained("deepseek-ai/DeepSeek-V4-Flash", torch_dtype=torch.bfloat16, trust_remote_code=True)
model.save_pretrained("/tmp/dsv4_resaved")
# Compare /tmp/dsv4_resaved/model-*.safetensors dtypes vs source
# Result: 417 keys that were FP32 in source are now BF16 in resaved
```

## Workaround (working in production)

Postprocess: after `save_pretrained` returns, walk the saved safetensors shards, read FP32 versions of the affected keys from the BF16 source checkpoint, and write them back in place (atomic per-shard via `.tmp + os.replace`). Working example: [`scripts/fixup_artifact.py`](https://github.com/canada-quant/dsv4-flash-w4a16-fp8-mtp/blob/main/scripts/fixup_artifact.py) in `canada-quant/dsv4-flash-w4a16-fp8-mtp`.

## Suggested fix

`save_pretrained` should preserve per-tensor dtype rather than coerce to model's `torch_dtype`. The current behavior is right for most weights (you wanted BF16 for the W4A16 main model) but wrong for plumbing tensors that ship FP32 for numerical-stability reasons. A whitelist of "always preserve source dtype" patterns per architecture would work; for DSv4-Flash the regex is roughly `r".*\.(hc_|attn_sink|ffn\.gate\.bias|compressor\.ape|indexer\.compressor\.ape)$"`.

Alternatively, a per-parameter dtype hint in the saved metadata + opt-in flag on the model config would solve it generically.

## Why it matters

Without the postprocess restore, the saved artifact has BF16 plumbing → numerical drift on the gating math + LM head logits → measurable quality regression. With the restore, our W4A16+FP8+MTP artifact at https://huggingface.co/canada-quant/DeepSeek-V4-Flash-W4A16-FP8-MTP hits 86.88% MMLU and 93.71% GSM8K (within SE of the un-restored-baseline-impossible-to-produce target). The sibling NVFP4 artifact at canada-quant/DeepSeek-V4-Flash-NVFP4-FP8-MTP applies the same restore postprocess.

This is filed by the canada-quant team during W4A16+FP8+MTP quantization work. See [`FINDINGS_FOR_SIBLING.md`](https://github.com/canada-quant/dsv4-flash-w4a16-fp8-mtp/blob/main/FINDINGS_FOR_SIBLING.md) §C13 for the diagnosis trace.

**Comment by ArjunSrivastava1** [comment]
got it, understand most of it, but above my paygrade

cc: @ArthurZucker, seems like a job for u

also mentioning this in the fix pr for easier tracking: #45892

**Comment by ArthurZucker** [comment]
Hey! Thanks, this does seem like an issue, thanks for opening! 

> A whitelist of "always preserve source dtype" patterns per architecture would work; for DSv4-Flash the regex is roughly r".*\.(hc_|attn_sink|ffn\.gate\.bias|compressor\.ape|indexer\.compressor\.ape)$".

this already exists and that's the `keep_in_fp32` attribute! Will check it out

**Comment by pasta-paul** [comment]
> **Disclosure:** this reply was generated with AI assistance (per the in-thread automated-agent disclosure request). The original issue body was also AI-assisted; humans reviewed before posting.

Thanks @ArthurZucker — confirming `keep_in_fp32` is the right lever here.

Locally we worked around this by clearing `_keep_in_fp32_modules_strict` during load (so MTP/conversion-mapping plumbing wouldn't fight us) and then patching the FP32 tensors back in after `save_pretrained`. That's a bandaid; the architecturally clean fix is to populate the DSv4 model class's `keep_in_fp32` with the patterns DeepSeek ships as FP32 in `deepseek-ai/DeepSeek-V4-Flash`:

- `.*\.hc_attn_(base|fn|scale)$`
- `.*\.hc_ffn_(base|fn|scale)$`
- `^(model|mtp\.0)\.hc_head_(base|fn|scale)$`
- `.*\.attn\.attn_sink$`
- `^layers\.\d+\.ffn\.gate\.bias$` (was `e_score_correction_bias`)
- `^layers\.\d+\.attn\.compressor\.ape$` (was `position_bias`)
- `^layers\.\d+\.attn\.indexer\.compressor\.ape$` (was `position_bias`)

417 tensors total in V4-Flash (44 layers + MTP). Same set applies to V4-Pro modulo layer count.

Happy to send a PR wiring this into `DeepseekV4PreTrainedModel.keep_in_fp32` if useful — let me know if you'd rather take it directly.
